src/program.rs

```diff
@@ -401,70 +405,178 @@ impl From<ColorOpt> for ColorChoice {
+            }
+        }
+    }
+
+    #[derive(Clone, Eq, PartialEq, Debug)]
+    pub struct ParseFormatError(String);
+
+    impl fmt::Display for ParseFormatError {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "unsupported export format {}", self.0)
+        }
+    }
+
+    impl std::str::FromStr for DocFormat {
+        type Err = ParseFormatError;
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            match s.to_lowercase().as_ref() {
+                "json" => Ok(DocFormat::Json),
+                "markdown" => Ok(DocFormat::Markdown),
+                _ => Err(ParseFormatError(String::from(s))),
+            }
+        }
+    }
+
+    #[derive(Clone, Debug, Serialize, Deserialize)]
+    #[serde(transparent)]
+    struct ExtractedDocumentation {
+        fields: HashMap<String, DocumentationField>,
+    }
+
+    #[derive(Clone, Debug, Serialize, Deserialize)]
+    struct DocumentationField {
+        /// Field value `DocOutput`s, if any
+        fields: Option<ExtractedDocumentation>,
+        /// Rendered type annotation, if any
+        types: Option<String>,
+        /// Rendered contract annotations
+        contracts: Vec<String>,
+        /// Rendered documentation, if any
+        documentation: Option<String>,
+    }
+
+    fn extract_documentation(rt: &RichTerm) -> Option<ExtractedDocumentation> {
+        match rt.term.as_ref() {
+            Term::Record(record) | Term::RecRecord(record, _, _) => {
+                let fields = record
+                    .fields
+                    .iter()
+                    .map(|(ident, field)| {
+                        (
+                            ident.label().to_owned(),
+                            DocumentationField {
+                                fields: field.value.as_ref().and_then(extract_documentation),
+                                types: field
+                                    .metadata
+                                    .annotation
+                                    .types
+                                    .as_ref()
+                                    .map(|lt| lt.types.to_string()),
+                                contracts: field
+                                    .metadata
+                                    .annotation
+                                    .contracts
+                                    .iter()
+                                    .map(|lt| lt.types.to_string())
+                                    .collect(),
+                                documentation: field.metadata.doc.clone(),
+                            },
+                        )
+                    })
+                    .collect();
+                Some(ExtractedDocumentation { fields })
+            }
+            _ => None,
+        }
+    }
+
+    /// Create a markdown or json file with documentation for the specified FileId.
     pub fn output_doc(
```

Comment:
I wonder if we can reorganize those free-standing functions as proper, smaller and composable methods, now that you've introduced a bunch of types. Something like `ExtractedDocumentation::extract_from_term(t: RichTerm) -> Self`, then `ExtractedDocumentation::extract_from_file(...) -> Self`, and two functions `to_markdown` and `to_json` (doing so we can also probably make all those functions independent of a free-standing `format` parameter).

Reply:
I've tried to do that refactoring. It does seem to make the code a bit more composable 👍